website/blog/2020-02-24-presto-lambda.md

```diff
@@ -0,0 +1,193 @@
+- Lambda capture has to be performed per invocation, as different value will be captured for each row).
+- However, we should use the same class representing captured lambda for every `BIND` call, otherwise we will generate too many classes.
+
+We use the same way as Java handles lambda and capture [3, 4]:
+- Lambda is represented as an object whose type is an functional interface (a.k.a. Single Abstract Method class)
+- The `invokedynamic` instruction is used to perform lambda capture:
+    * During the first `invokedynamic` call, the class represents captured lambda (which is a functional interface) and the method that performs capture will be generated. This step is also called linkage and will only be done once.
+    * Every `invokedynamic` call performs capture and returns an instance of the desired functional interface. 
+
+With this design, `BIND` function will always be fused together with the lambda generation step to generate captured lambda in one single step -- we cannot first generate an object that represents uncaptured lambda, and performs a separate partial application step. Note this implementation also doesn't allow more general higher-order function that returns a function as result.
+
+## Lambda in Aggregation 
+
+While lambda is originally introduced to help query nested data with scalar functions, we also noted it can be used in aggregation functions to allow more flexible analytics. The initial preliminary support of lambda in aggregation is added in [Presto#12084](https://github.com/prestodb/presto/pull/12084), with a [reduce_agg](https://prestodb.io/docs/current/functions/aggregate.html#reduce_agg) function for demonstration purpose. By making the input and combine function to be specified with lambda, this function *conceptually* allows to write User-Defined Aggregation Function (UDAF). The following example shows how to use `reduce_agg` to compute group-wise product (instead of sum):
+
+```sql
+SELECT id, 
+reduce_agg(value, 1, (a, b) -> a * b, (a, b) -> a * b) prod
+FROM (
+    VALUES
+        (1, 2),
+        (1, 3),
+        (1, 4),
+        (2, 20),
+        (2, 30),
+        (2, 40)
+) AS t(id, value)
+GROUP BY id;
+
+ id | prod
+----+-------
+  2 | 24000
+  1 |    24
+(2 rows)
+```
+
+Unfortunately, due to [JDK-8017163](https://bugs.openjdk.java.net/browse/JDK-8017163), aggregation state with `Slice` or `Block` as native container type is intentionally not supported yet, as it may result in excessive JVM memory usage of remembered set (see [Presto#9553](https://github.com/prestodb/presto/issues/9553) for details). This makes this function not yet practically useful. Once [JDK-8017163](https://bugs.openjdk.java.net/browse/JDK-8017163) is fixed in later version of JVM, we are looking forward to enabling it with more general types to allow more flexible analytics in aggregations!
```

Comment:
This is another case where the reader probably wants to know why it doesn't work rather than linking to the GH issue.

Reply:
"in later versions of the JVM. We  are looking forward"

Reply:
@aweisberg : The previous clause starts with "once", so we cannot end it with period right? ;)

Reply:
You are right.

`Once [JDK-8017163](https://bugs.openjdk.java.net/browse/JDK-8017163) is fixed we can enable lambdas in aggregations with more general types allowing more flexible analytics.`